sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst

```diff
@@ -0,0 +1,162 @@
+C and C++ Standard libraries support
+===================================
+
+This extension enables a set of functions from the C and C++ standard
+libraries, and allows to use them in SYCL device code.
+
+Function declarations are taken from the standard headers (e.g. from
+<assert.h> or <complex>), and the corresponding header has to be
+explicitly included in user code.
+
+Implementation requires a special device library to be linked with a
+SYCL program. The library should match the C or C++ standard library
+used to compile the program:
+
+For example, on Linux with GNU glibc:
+.. code:
+   clang++ -fsycl -c main.cpp -o main.o
+   clang++ -fsycl main.o $(SYCL_INSTALL)/lib/libsycl-glibc.o -o a.out
+
+or, in case of Windows:
+.. code:
+   clang++ -fsycl -c main.cpp -o main.obj
+   clang++ -fsycl main.obj %SYCL_INSTALL%/lib/libsycl-msvc.o -o a.exe
+
+List of supported functions from C standard library:
+  - assert macro (from assert.h)
+
+NOTE: only the GNU glibc and Microsoft C libraries are currently
+supported.
+
+Example of usage
+================
+
+.. code: c++
+   #include <assert.h>
+   #include <CL/sycl.hpp>
+
+   template <typename T, size_t N>
+   void simple_vadd(const std::array<T, N>& VA, const std::array<T, N>& VB,
+                    std::array<T, N>& VC) {
+     // ...
+     cl::sycl::range<1> numOfItems{N};
+     cl::sycl::buffer<T, 1> bufferA(VA.data(), numOfItems);
+     cl::sycl::buffer<T, 1> bufferB(VB.data(), numOfItems);
+     cl::sycl::buffer<T, 1> bufferC(VC.data(), numOfItems);
+
+     deviceQueue.submit([&](cl::sycl::handler& cgh) {
+       auto accessorA = bufferA.template get_access<sycl_read>(cgh);
+       auto accessorB = bufferB.template get_access<sycl_read>(cgh);
+       auto accessorC = bufferC.template get_access<sycl_write>(cgh);
+
+       cgh.parallel_for<class SimpleVadd<T>>(numOfItems,
+       [=](cl::sycl::id<1> wiID) {
+           accessorC[wiID] = accessorA[wiID] + accessorB[wiID];
+           assert(accessorC[wiID] > 0 && "Invalid value");
```

Comment:
> Is assert a good example? It relies on some kernel features not available in OpenCL...

Right, assert is an edge case and it is probably a bad example.
However, it is the only standard C and C++ function supported at this moment.

> Are there real use cases?

For assert? It is widely used as a sanity check in host code, so I don't see a reason why it cannot be used for device code as well.

> What about this implementation:
> 
> ```
> #ifdef __SYCL_DEVICE_ONLY__
> #define NDEBUG 1
> #endif 
> ```

That should work if you define the macro before `#include <assert.h>`.
Beware any significant changes under `#if NDEBUG`:
```
struct shared {
  int i;
#ifndef NDEBUG
  int debug;
#endif
};
```

Reply:
I am unsure that `assert()` is really "supported", in the sense that the behaviour of what is proposed here is undefined.
 
The macro `__SYCL_DEVICE_ONLY__` is set by the SYCL device compiler.
Having the device compiler driver to also set `NDEBUG` would be abusive but would somehow solve a little bit of the "problem".
But you could always `#define assert()` in your `sycl.hpp` even after including "cassert".
Anyway all this looks messy because, well, this is outside of the scope of SYCL and kernels...

Reply:
> Having the device compiler driver to also set `NDEBUG` would be abusive but would somehow solve a little bit of the "problem".
> But you could always `#define assert()` in your `sycl.hpp` even after including "cassert".

There are two problems:
1. This solution depends on order of `#include` directives: if you include `assert.h` after `sycl.hpp`, the macro will be redefined again.
```
#ifdef  _ASSERT_H

# undef _ASSERT_H
# undef assert
[...]
#endif /* assert.h      */
```

2. Even if you manage to define the macro properly, there is still a problem of not having a proper abort function.

So we avoid this dance around `__assert_fail` and `_wassert`, but we still have to manage the devicelib assert extension.

> Anyway all this looks messy because, well, this is outside of the scope of SYCL and kernels...

Agree.